### PR TITLE
Implementation of MSC3824 to add action= param on SSO login

### PR DESCRIPTION
**Pull request by hughns**
---
Here's what your changelog entry will look like:

## ✨ Features
 * Implementation of MSC3824 to add action= param on SSO login ([\#2398](https://github.com/matrix-org/matrix-js-sdk/pull/2398)). Contributed by @hughns.

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -876,6 +877,8 @@ export type ClientEventHandlerMap = {
     & HttpApiEventHandlerMap
     & BeaconEventHandlerMap;
 
+const UNSTABLE_MSC3824_ACTION_PARAM = new UnstableValue("action", "org.matrix.msc3824.action");
```
Comment (turt2live) [inline src/client.ts:880]: we typically just call it something like `OIDC_ACTION_PARAM` so we can change the type to a `NamespacedValue` without breaking the SDK itself.

Comment (hughns) [reply src/client.ts]: I've renamed it `SSO_ACTION_PARAM`.

File: src/client.ts [hunk 2]
```diff
@@ -7143,15 +7146,24 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
...
      * @param {string} idpId The ID of the Identity Provider being targeted, optional.
+     * @param {"login"|"register"} action the SSO flow to indicate to the IdP, optional.
      * @return {string} The HS URL to hit to begin the SSO login process.
      */
-    public getSsoLoginUrl(redirectUrl: string, loginType = "sso", idpId?: string): string {
+    public getSsoLoginUrl(
+        redirectUrl: string,
+        loginType = "sso",
+        idpId?: string,
+        action?: "login" | "register",
```
Comment (SimonBrandner) [inline src/client.ts:7156]: Would we benefit from this being an enum?

File: src/client.ts [hunk 3]
```diff
@@ -7143,15 +7146,24 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
...
+    ): string {
         let url = "/login/" + loginType + "/redirect";
         if (idpId) {
             url += "/" + idpId;
         }
 
-        return this.http.getUrl(url, { redirectUrl }, PREFIX_R0);
+        const params = { redirectUrl };
+        params[SSO_ACTION_PARAM.unstable] = action;
```
Comment (SimonBrandner) [inline src/client.ts:7164]:
```suggestion
        const params = { 
            redirectUrl,
            [SSO_ACTION_PARAM.unstable]: action,
        };
```

I think this looks a bit nicer

<hr>

Also, are we ok with sending the server `undefined`s?

**Review by SimonBrandner — APPROVED** [review]

**Review by weeman1337 — CHANGES_REQUESTED** [review]
@hughns can you add a test for it?

File: src/client.ts (hunk 2 above)
Comment (hughns) [reply src/client.ts]: I've made it an enum.

**Comment by hughns** [comment]
> @hughns can you add a test for it?

I've added some some simple test cases.

**Review by weeman1337 — APPROVED** [review]
:+1:

**Pull request merged**